File: javaTool/src/main/java/nl/um/cds/triplifier/rdf/ontology/DBO.java
package nl.um.cds.triplifier.rdf.ontology;

import org.eclipse.rdf4j.model.IRI;
import org.eclipse.rdf4j.model.impl.SimpleValueFactory;
import org.eclipse.rdf4j.model.vocabulary.OWL;
import org.eclipse.rdf4j.model.vocabulary.RDF;
import org.eclipse.rdf4j.model.vocabulary.RDFS;
import org.eclipse.rdf4j.repository.RepositoryConnection;

public class DBO {
    public static final String NAMESPACE = "http://um-cds/ontologies/databaseontology/";
    public static final IRI DATABASETABLE = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "TableRow");
    public static final IRI DATABASECOLUMN = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "ColumnCell");
    public static final IRI DATABASECELL = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "Cell");
    public static final IRI PRIMARYKEY = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "PrimaryKey");
    public static final IRI FOREIGNKEY = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "ForeignKey");

    public static final IRI HAS_COLUMN = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "has_column");
    public static final IRI HAS_VALUE = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "has_value");
    public static final IRI HAS_CELL = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "has_cell");
    public static final IRI HAS_UNIT = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "has_unit");
    // public static final IRI COLUMNREFERENCE = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "ColumnReference");
    public static final IRI HAS_TARGET_COLUMN = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "has_target_column");
    public static final IRI FK_REFERS_TO = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "fk_refers_to");

    public static final IRI TABLE = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "table");
    public static final IRI CATALOG = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "catalog");
    public static final IRI SCHEMA = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "schema");
    public static final IRI COLUMN = SimpleValueFactory.getInstance().createIRI(NAMESPACE, "column");

    public static void addOntologyToDatabaseConnection(RepositoryConnection conn) {
        conn.setNamespace("dbo", NAMESPACE);

        conn.add(DATABASETABLE, RDF.TYPE, OWL.CLASS);
        conn.add(DATABASECOLUMN, RDF.TYPE, OWL.CLASS);
        conn.add(PRIMARYKEY, RDF.TYPE, OWL.CLASS);
        conn.add(PRIMARYKEY, RDFS.SUBCLASSOF, DATABASECOLUMN);
        conn.add(FOREIGNKEY, RDF.TYPE, OWL.CLASS);
        conn.add(FOREIGNKEY, RDFS.SUBCLASSOF, DATABASECOLUMN);

        conn.add(HAS_COLUMN, RDF.TYPE, OWL.OBJECTPROPERTY);
        conn.add(HAS_COLUMN, RDFS.DOMAIN, DATABASETABLE);
        conn.add(HAS_COLUMN, RDFS.RANGE, DATABASECOLUMN);

        conn.add(HAS_VALUE, RDF.TYPE, OWL.OBJECTPROPERTY);
        conn.add(HAS_VALUE, RDFS.DOMAIN, DATABASECOLUMN);
        conn.add(HAS_VALUE, RDFS.RANGE, DATABASECELL);

        conn.add(HAS_CELL, RDF.TYPE, OWL.DATATYPEPROPERTY);
        conn.add(HAS_CELL, RDFS.DOMAIN, DATABASECOLUMN);
        conn.add(HAS_UNIT, RDF.TYPE, OWL.ANNOTATIONPROPERTY);
        conn.add(HAS_UNIT, RDFS.DOMAIN, DATABASECOLUMN);

        conn.add(HAS_TARGET_COLUMN, RDF.TYPE, OWL.ANNOTATEDPROPERTY);
        conn.add(HAS_TARGET_COLUMN, RDFS.DOMAIN, DATABASECOLUMN);

        conn.add(FK_REFERS_TO, RDF.TYPE, OWL.OBJECTPROPERTY);
        conn.add(FK_REFERS_TO, RDFS.LABEL, SimpleValueFactory.getInstance().createLiteral("foreign key refers to"));
        conn.add(FK_REFERS_TO, RDFS.DOMAIN, DATABASECOLUMN);
        conn.add(FK_REFERS_TO, RDFS.RANGE, PRIMARYKEY);

        conn.add(TABLE, RDF.TYPE, OWL.ANNOTATIONPROPERTY);
        conn.add(TABLE, RDFS.RANGE, DATABASETABLE);
        conn.add(TABLE, RDFS.DOMAIN, DATABASECOLUMN);
        conn.add(CATALOG, RDF.TYPE, OWL.ANNOTATIONPROPERTY);
        conn.add(CATALOG, RDFS.DOMAIN, DATABASETABLE);
        conn.add(SCHEMA, RDF.TYPE, OWL.ANNOTATIONPROPERTY);
        conn.add(SCHEMA, RDFS.DOMAIN, DATABASETABLE);
        conn.add(COLUMN, RDF.TYPE, OWL.ANNOTATIONPROPERTY);
        conn.add(COLUMN, RDFS.DOMAIN, DATABASECOLUMN);
    }

}

File: pythonTool/rdf/ontology/dbo.py
from rdflib import Namespace, RDF, RDFS, OWL, Literal

DBO = Namespace("http://um-cds/ontologies/databaseontology/")

DATABASETABLE = DBO.TableRow
DATABASECOLUMN = DBO.ColumnCell
DATABASECELL = DBO.Cell
PRIMARYKEY = DBO.PrimaryKey
FOREIGNKEY = DBO.ForeignKey

HAS_COLUMN = DBO.has_column
HAS_VALUE = DBO.has_value
HAS_CELL = DBO.has_cell
HAS_UNIT = DBO.has_unit
HAS_TARGET_COLUMN = DBO.has_target_column
FK_REFERS_TO = DBO.fk_refers_to

TABLE = DBO.table
CATALOG = DBO.catalog
SCHEMA = DBO.schema
COLUMN = DBO.column


def add_ontology_to_graph(graph):
    graph.bind("dbo", DBO)
    graph.add((DATABASETABLE, RDF.type, OWL.Class))
    graph.add((DATABASECOLUMN, RDF.type, OWL.Class))
    graph.add((DATABASECELL, RDF.type, OWL.Class))
    graph.add((PRIMARYKEY, RDF.type, OWL.Class))
    graph.add((PRIMARYKEY, RDFS.subClassOf, DATABASECOLUMN))
    graph.add((FOREIGNKEY, RDF.type, OWL.Class))
    graph.add((FOREIGNKEY, RDFS.subClassOf, DATABASECOLUMN))

    graph.add((HAS_COLUMN, RDF.type, OWL.ObjectProperty))
    graph.add((HAS_COLUMN, RDFS.domain, DATABASETABLE))
    graph.add((HAS_COLUMN, RDFS.range, DATABASECOLUMN))

    graph.add((HAS_VALUE, RDF.type, OWL.ObjectProperty))
    graph.add((HAS_VALUE, RDFS.domain, DATABASECOLUMN))
    graph.add((HAS_VALUE, RDFS.range, DATABASECELL))

    graph.add((HAS_CELL, RDF.type, OWL.DatatypeProperty))
    graph.add((HAS_CELL, RDFS.domain, DATABASECOLUMN))
    graph.add((HAS_UNIT, RDF.type, OWL.AnnotationProperty))
    graph.add((HAS_UNIT, RDFS.domain, DATABASECOLUMN))

    graph.add((HAS_TARGET_COLUMN, RDF.type, OWL.AnnotationProperty))
    graph.add((HAS_TARGET_COLUMN, RDFS.domain, DATABASECOLUMN))

    graph.add((FK_REFERS_TO, RDF.type, OWL.ObjectProperty))
    graph.add((FK_REFERS_TO, RDFS.label, Literal("foreign key refers to")))
    graph.add((FK_REFERS_TO, RDFS.domain, DATABASECOLUMN))
    graph.add((FK_REFERS_TO, RDFS.range, PRIMARYKEY))

    graph.add((TABLE, RDF.type, OWL.AnnotationProperty))
    graph.add((TABLE, RDFS.range, DATABASETABLE))
    graph.add((TABLE, RDFS.domain, DATABASECOLUMN))
    graph.add((CATALOG, RDF.type, OWL.AnnotationProperty))
    graph.add((CATALOG, RDFS.domain, DATABASETABLE))
    graph.add((SCHEMA, RDF.type, OWL.AnnotationProperty))
    graph.add((SCHEMA, RDFS.domain, DATABASETABLE))
    graph.add((COLUMN, RDF.type, OWL.AnnotationProperty))
    graph.add((COLUMN, RDFS.domain, DATABASECOLUMN))
